### ISSUE TITLE
Spring Boot 3.x.x update throws JedisClusterOperationException Could not initialize cluster slots cache.

### ISSUE THREAD

**Issue by kableprime** [redis/jedis#3392]
### Actual behavior

We have a spring boot application using jedis for connecting to redis in some of the services but since we updated to spring boot 3.x.x all unit tests using jedis fail with the following stack:
(I have shortened the entire stack, because the redis config and redis method is written in a our ownlibrary that we re-use across multiple apps)

```
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'redisUtils' defined in URL [jar:file:/Users/I343767/.m2/repository/com/sap/etd/cloud/streaming-commons/3.2.5/streaming-commons-3.2.5.jar!/com/sap/etd/cloud/commons/utils/RedisUtils.class]: Unsatisfied dependency expressed through constructor parameter 1: Error creating bean with name 'getRedisTemplate' defined in class path resource [com/sap/etd/cloud/commons/config/redis/RedisConfig.class]: Failed to instantiate [org.springframework.data.redis.core.StringRedisTemplate]: Factory method 'getRedisTemplate' threw exception with message: Error creating bean with name 'getClusterRedisConnectionFactory' defined in class path resource [com/sap/etd/cloud/commons/config/redis/RedisConfig.class]: Could not initialize cluster slots cache.
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:245)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1352)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1189)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1417)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1337)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
	... 103 more
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'getRedisTemplate' defined in class path resource [com/sap/etd/cloud/commons/config/redis/RedisConfig.class]: Failed to instantiate [org.springframework.data.redis.core.StringRedisTemplate]: Factory method 'getRedisTemplate' threw exception with message: Error creating bean with name 'getClusterRedisConnectionFactory' defined in class path resource [com/sap/etd/cloud/commons/config/redis/RedisConfig.class]: Could not initialize cluster slots cache.
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:659)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:493)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1332)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1162)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1417)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1337)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
	... 117 more
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.data.redis.core.StringRedisTemplate]: Factory method 'getRedisTemplate' threw exception with message: Error creating bean with name 'getClusterRedisConnectionFactory' defined in class path resource [com/sap/etd/cloud/commons/config/redis/RedisConfig.class]: Could not initialize cluster slots cache.
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:171)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:655)
	... 131 more
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'getClusterRedisConnectionFactory' defined in class path resource [com/sap/etd/cloud/commons/config/redis/RedisConfig.class]: Could not initialize cluster slots cache.
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:598)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:225)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1310)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1271)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveBean(DefaultListableBeanFactory.java:484)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:339)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:332)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1150)
	at com.sap.etd.cloud.commons.config.redis.RedisConfig.getRedisTemplate(RedisConfig.java:71)
	at com.sap.etd.cloud.commons.config.redis.RedisConfig$$SpringCGLIB$$0.CGLIB$getRedisTemplate$0(<generated>)
	at com.sap.etd.cloud.commons.config.redis.RedisConfig$$SpringCGLIB$$2.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
	at com.sap.etd.cloud.commons.config.redis.RedisConfig$$SpringCGLIB$$0.getRedisTemplate(<generated>)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:139)
	... 132 more
Caused by: redis.clients.jedis.exceptions.JedisClusterOperationException: Could not initialize cluster slots cache.
	at redis.clients.jedis.providers.ClusterConnectionProvider.initializeSlotsCache(ClusterConnectionProvider.java:56)
	at redis.clients.jedis.providers.ClusterConnectionProvider.<init>(ClusterConnectionProvider.java:32)
	at redis.clients.jedis.UnifiedJedis.<init>(UnifiedJedis.java:130)
	at redis.clients.jedis.JedisCluster.<init>(JedisCluster.java:179)
	at redis.clients.jedis.JedisCluster.<init>(JedisCluster.java:172)
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.createCluster(JedisConnectionFactory.java:413)
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.createCluster(JedisConnectionFactory.java:377)
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.afterPropertiesSet(JedisConnectionFactory.java:300)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1816)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1766)
	... 155 more
	Suppressed: redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.
		at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:63)
		at redis.clients.jedis.DefaultJedisSocketFactory.createSocket(DefaultJedisSocketFactory.java:87)
		at redis.clients.jedis.Connection.connect(Connection.java:190)
		at redis.clients.jedis.Connection.initializeFromClientConfig(Connection.java:348)
		at redis.clients.jedis.Connection.<init>(Connection.java:52)
		at redis.clients.jedis.providers.ClusterConnectionProvider.initializeSlotsCache(ClusterConnectionProvider.java:45)
		... 164 more
		Suppressed: java.net.ConnectException: Connection refused: 127.0.0.1:6379
			at java.base/sun.nio.ch.Net.pollConnect(Native Method)
			at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
			at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:549)
			at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
			at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
			at java.base/java.net.Socket.connect(Socket.java:633)
			at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:73)
			... 169 more
		Suppressed: java.net.ConnectException: Connection refused: 0:0:0:0:0:0:0:1:6379
			at java.base/sun.nio.ch.Net.pollConnect(Native Method)
			at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
			at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:549)
			at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
			at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
			at java.base/java.net.Socket.connect(Socket.java:633)
			at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:73)
			... 169 more
```

### Redis / Jedis Configuration

This is the RedisConfig class:
```
@Configuration
@EnableAsync
@Log4j2
public class RedisConfig {
	
	private final ApplicationContext context;
	
	private final CfEnvUtils cfEnvUtils;
	
	@Autowired
	public RedisConfig(ApplicationContext context, CfEnvUtils cfEnvUtils) {
		this.context = context;
		this.cfEnvUtils = cfEnvUtils;
	}
	
	
	private JedisClientConfiguration getJedisClientConfiguration() {
		log.info("Preparing Jedis Client Config for ");
		return JedisClientConfiguration.builder().useSsl().build();
	}
	
	
	@Bean
	public RedisConnectionFactory getClusterRedisConnectionFactory() {
		if (cfEnvUtils.isInCloud()) {
			CfEnv cfEnv = new CfEnv();
			String redisHost = cfEnv.findCredentialsByTag(REDIS_TAG).getHost();
			String redisPort = cfEnv.findCredentialsByTag(REDIS_TAG).getPort();
			String redisPassword = cfEnv.findCredentialsByTag(REDIS_TAG).getPassword();
			JedisConnectionFactory jcf;
			JedisClientConfiguration clientConfig = getJedisClientConfiguration();
			RedisClusterConfiguration clusterConfiguration = new RedisClusterConfiguration(
					Collections.singletonList(redisHost + ":" + redisPort));
			clusterConfiguration.setPassword(redisPassword);
			jcf = new JedisConnectionFactory(clusterConfiguration, clientConfig);
			return jcf;
		} else {
			JedisClientConfiguration clientConfig = getJedisClientConfiguration();
			RedisClusterConfiguration clusterConfiguration = new RedisClusterConfiguration(
					Collections.singletonList("localhost:" + 6379));
			JedisConnectionFactory jcf;
			jcf = new JedisConnectionFactory(clusterConfiguration, clientConfig);
			return jcf;
		}
		
	}
	
	
	@Bean
	public StringRedisTemplate getRedisTemplate() {
		StringRedisTemplate stringRedisTemplate = new StringRedisTemplate(context.getBean(RedisConnectionFactory.class));
		stringRedisTemplate.setEnableTransactionSupport(true);
		return stringRedisTemplate;
	}
	
}
```

### Steps to reproduce:

1. Upgrade to spring 3.0.6
2. Run unit tests that use jedis and they fail

#### Java version: 17

**UPDATE** Debugging JedisConnectionException results: "Failed to connect to any host resolved for DNS name"
Is there any change in the latest jedis versions that require different redis configurations?

**Comment by sazzad16** [comment]
I wonder what is in your unit tests. Could it be just limited to creating a cluster connection object? Without executing any other command?

Why I'm wondering this because of a certain difference between Jedis 3.x and Jedis 4+ which seems to matter in your case.
In Jedis 3.x, when a JedisCluster object was being created with all cluster nodes down/unavailable, the constructor would not throw any exception and the object would be created BUT it would be a stale object with which no server operations can be done.
To address requests from users, Jedis 4.0 started throwing exception in this case with message `Could not initialize cluster slots cache.`.

I hope I have been able to identify your problem properly and you have found it helpful. Thanks.

**Comment by kableprime** [comment]
I think you perfectly pointed the issue, I do not use any redis operations in the unit test but the unit test needs to load the entire spring application context. 
And before Jedis 4 the exception just wasn't being thrown.

I am not sure how to handle this situation because rolling back to jedis 3.x is not possible because there are incompatibilities between spring-data-redis and jedis so I think I need to provide the redis server in all unit test environments. Do you have any other workarounds in mind? @sazzad16

**Comment by sazzad16** [comment]
Can you put a try-catch block at the concerned place?
Perhaps returning a `null` instead of a JedisCluster object, if necessary.

**Comment by nikhilgurjar** [comment]
@sazzad16  issue does not come when returning jedisConnectionfactory. issue is happening after that. Can we override JedisClusterOperationException so that it just logs the error and not make whole application to  shut down

**Comment by sazzad16** [comment]
@nikhilgurjar 

1. Are you and the OP (**kableprime**) are from same team? (Just to get more information about the user group.)
2. To clarify, your preference is to have an inoperable object (cluster) for the entirety of your application (instead of getting the error). Right?
3. (2) is quite opposite to what some of our users want. It'll be difficult to choose one user group over the others.
4. Can't you create a `Connection`, send a `ping` and do something else when the connection is unavailable?
5. Does a resolution through a system property work for you?

**Comment by nikhilgurjar** [comment]
1. No we are not from same team
2. Can’t you create a flag where if wanted to throw error we can make that flag = true in intializeslotcache class 
3. Which system properly are you talking about, can you tell me that 
4. And can we create some dummy connection if ping fails as you told in point 4?

**Comment by nikhilgurjar** [comment]
when a JedisCluster object was being created with all cluster nodes down/unavailable, the constructor would not throw any exception and the object would be created BUT it would be a stale object with which no server operations can be done.

can we do custom implement this in jedis 4 anyhow without changing internal implementation ?

**Comment by sazzad16** [comment]
> Which system properly are you talking about, can you tell me that

There's no such thing available ATM. I asked to think about a solution using a **new** system property.

**Comment by poorva-swapstech** [comment]
Facing the same error into my application, did we have any resolution for this one ?

**Comment by sazzad16** [comment]
> Facing the same error into my application, did we have any resolution for this one ?

@poorva-swapstech https://github.com/redis/jedis/wiki/FAQ#how-to-avoid-cluster-initialization-error

**Comment by poorva-swapstech** [comment]
> > Facing the same error into my application, did we have any resolution for this one ?
> 
> @poorva-swapstech https://github.com/redis/jedis/wiki/FAQ#how-to-avoid-cluster-initialization-error

@sazzad16 - This solution did not help me. Also, just some more information for your reference - 
- Facing this issue in my local (not yet tested in higher environments)
- Jedis version is - **4.3.2**
- spring-data-redis version is **3.1.2**

**Comment by sazzad16** [comment]
> Facing this issue in my local

I don't have any knowledge about the configuration and the state of your local

> Jedis version is - 4.3.2

It is available since Jedis 4.4.2

@poorva-swapstech
